### PR TITLE
Commit in metadata

### PR DESCRIPTION
**Pull request by bitdivine**
# Motivation
Requests from the nns & testing teams to publish the commit of our canisters.  In fact, it is becoming common to publish both the commit and the API did file so we cover both on both canisters.  Four with one stroke.

# Changes
* Created a new command to add metadata.  This is NOT in the canister build command so that the canister build can be cached, and commits that do not affect a canister wasm don't cause a rebuild just because the commit has changed.
* Created a test for the command
* Called the command in the dockerfile
* Created a test for deployments from the docker build

# Tests
* Tests are included.  See CI for the results.

### CODE REVIEW

File: .github/workflows/aggregator.yaml [hunk 1]
```diff
@@ -24,6 +24,7 @@ jobs:
           file: Dockerfile
           build-args: |
             DFX_NETWORK=local
+            COMMIT=${{ github.sha }}
```
Comment (bitdivine) [inline .github/workflows/aggregator.yaml:27]: We need the git commit inside docker but  don't pull .git into docker.  Instead we pass in the commit to every docker  build.

Alternative experimented with:  Pull .git in to a dedicated stage, read the commit and throw away everything else from that stage.  That fails for git worktrees and will also fail for anyone trying to build from a release, that contains a tarball of the git content but not `.git/HEAD`.  I don't know that anyone is building from releases but I do know that at least I am using git worktrees and I would rather not break my build environment, even if that means making this PR bigger.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -83,6 +84,10 @@ jobs:
             # grep ... fails if the value is implausibly short.
             key="$key" jq -re '.[0][env.key]' nns_dapp_args_in_page.json | grep -E ...
           done
+      - name: Install ic-wasm
+        run: cargo install "ic-wasm@$(jq -r '.defaults.build.config.IC_WASM_VERSION' dfx.json)"
```
Comment (bitdivine) [inline .github/workflows/build.yml:88]: `ic-wasm` is a tool used to edit canister wasms.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -83,6 +84,10 @@ jobs:
...
+      - name: Check that metadata is present
+        run: scripts/dfx-wasm-metadata-add.test
```
Comment (bitdivine) [inline .github/workflows/build.yml:90]: This tests that the metadata is added to the wasms.

File: scripts/dfx-wasm-metadata-add [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  echo "ERROR: There should be a wasm file at: '$DFX_CANISTER_WASM'"
+  exit 1
+} >&2
+
+: "Find the candid file"
+test -n "${DFX_CANISTER_DID:-}" || DFX_CANISTER_DID="$(jq -r '.canisters[env.DFX_CANISTER_NAME].candid // ""' dfx.json)"
+test -n "${DFX_CANISTER_DID:-}" || {
+  echo "ERROR: Unable to determine the path to the candid file."
+  echo "       Please either:"
+  echo "         Use --wasm PATH_TO_WASM"
```
[the remark this answers is not in the document]
Comment (bitdivine) [reply scripts/dfx-wasm-metadata-add:44]: Yes, thank you.  Done.

**Review by dskloetd — APPROVED** [review]

**Pull request merged**